Finish main page hero

alexzhang1618 opened the pull request:
# Description

What changes did you make? List all distinct problems that this PR addresses. Explain any relevant
motivation or context.

- Implement the main page hero
- This also includes check-in handling through the QR code/link, and form on the main page

Things to do:
- cool and funny animations on the hero

# Type of Change

- [ ] Bug Fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as
      expected)
- [ ] Logistics Change (A change to a README, description, or dev workflow setup like
      linting/formatting)
- [ ] Continuous Integration Change (Related to deployment steps or continuous integration
      workflows)
- [ ] Other: (Fill In) <!-- Edit this type of change if you select this -->

# Testing

I have tested that my changes fully resolve the linked issue ...

- [x] locally on Desktop.
- [ ] locally on mobile - use https://ngrok.io to get a copy on a mobile device
- [ ] on the live deployment preview on Desktop.
- [ ] on the live deployment preview on Mobile.
- [ ] I have run and passed all new and existing Cypress tests. Add screenshots below.

# Checklist

- [x] I have performed a self-review of my own code.
- [x] I have followed the style guidelines of this project.
- [x] I have documented my code's `src/lib` functions and commented hard to understand areas
      anywhere else.
- [x] My changes produce no new warnings.

# Screenshots

Normal screens
<img width="1440" alt="Screenshot 2024-01-02 at 12 24 48 PM" src="https://github.com/acmucsd/membership-portal-ui-v2/assets/27360825/515f8f17-c085-4a52-a457-2eb6d544f7a2">
Wide screens
<img width="1165" alt="Screenshot 2024-01-02 at 12 25 07 PM" src="https://github.com/acmucsd/membership-portal-ui-v2/assets/27360825/f086eda6-ca8c-4e04-b127-5df55e81ace3">
Mobile screens
<img width="218" alt="Screenshot 2024-01-02 at 12 25 28 PM" src="https://github.com/acmucsd/membership-portal-ui-v2/assets/27360825/04a53e0e-69ec-4723-8f2d-631efb9560e0">

alexzhang1618 committed:
Finish main page hero

alexzhang1618 committed:
Cleanup CSS

alexzhang1618 committed:
Fix height bug :)

alexzhang1618 committed:
Add action to form so the ios keyboard will say go instead of return

farisashai reviewed (commented):
I really like how you linked every other page from the home page in a good way without needing to use the navbar. I honestly think we should have a "about acm" or "learn more" link in this home section and completely remove the link from the main navigation since it's so insignificant

farisashai commented on src/pages/index.tsx:
```diff
@@ -1,22 +1,73 @@
-const PortalHomePage = ({ pastEvents, upcomingEvents, liveEvents, attendances }: HomePageProps) => {
+const checkin = async (token: string, code: string) => {
+  try {
+    const response = await attendEvent(token, code);
```
Are we still following the convention of using the EventManager files to wrap api calls or scrapping that idea completely? I think it is a cleaner structure to isolate logic for the api arguments and error handling outside of page files so we can focus on the outputs only.

SheepTester commented on src/components/home/Hero/index.tsx:
```diff
@@ -0,0 +1,98 @@
+        alt="Landing page graphic"
+        priority
+        fill
+      />
```
you could also remove the waves from the raccoon image so you can keep waves.svg visible always, and it allows for the raccoon to move more independently if we want to

SheepTester commented on src/components/home/Hero/style.module.scss:
```diff
@@ -0,0 +1,154 @@
+
+        &:hover {
+          box-shadow: 0 5px 10px rgba(0, 0, 0, 0.3);
+        }
```
it's a bit weird having a hover effect for the check-in card since clicking it doesn't do anything

SheepTester commented on src/components/home/Hero/style.module.scss:
```diff
@@ -0,0 +1,154 @@
+      .checkin {
+        background-color: var(--theme-accent-line-2-transparent);
+        border-radius: 1rem;
+        box-shadow: 0 2px 5px rgba(0, 0, 0, 0.3);
```
we have `var(--theme-shadow)` which you could use for shadows

alexzhang1618 committed:
Rewrite checkin logic into EventManager

SheepTester commented on src/components/home/Hero/style.module.scss:
```diff
@@ -0,0 +1,154 @@
+    object-fit: cover;
+    object-position: 100% 100%;
+    // We have to manually add width since it's taking the width of the container, which is padded.
+    width: calc(100% + 4rem) !important;
```
you could avoid using `!important` here by instead applying negative margin on `.hero`:

```css
.hero {
  margin: 0 -4rem;
  padding: 0 4rem;
}
```

alexzhang1618 committed:
Stop submission of checkin code if empty

SheepTester commented on src/components/home/Hero/style.module.scss:
```diff
@@ -0,0 +1,154 @@
+    .checkin {
+      .checkinButtons {
+        .submit {
+          display: none !important;
```
if you want to avoid using `!important`, you can move these `@media` queries into the styles above, like

```scss
.submit {
  ...
  
  @media screen and (width <= vars.$breakpoint-sm) {
    display: none;
  }
}
```

so you don't have to fight CSS precedence rules. it also keeps the styles for the elements next to each other rather than at the end of the file

alexzhang1618 committed:
oops accidentally left a space

SheepTester commented on src/pages/index.tsx:
```diff
@@ -1,22 +1,68 @@
+        router.replace('/', undefined, { shallow: true });
+      });
+    }
+  }, [authToken, router]);
```
do you think this could be done on the server side, in `getServerSidePropsFunc`?

farisashai replied:
I would do the checkin code server side and return an event object that you use to immediately pop up the modal with the successfully checked into X

SheepTester commented on src/components/home/Hero/index.tsx:
```diff
@@ -0,0 +1,98 @@
+              {' membership points.'}
+            </Typography>
+          </div>
+        </div>
```
is it possible for a user to only have one point? (i.e. is it possible for someone to see "You have 1 membership points"?)

You should use `component="span"` instead of setting `display: inline` on divs. Alternatively, you could reduce code repetition by putting `Link` and `<strong>` inside one `Typography` element, like this:

```suggestion
        <div className={styles.header}>
          <Typography variant="display/light/large" className={styles.headline}>
            Welcome to ACM,&nbsp;
            <Link href={config.profileRoute}>
              <strong>
                {user.firstName}
              </strong>
            </Link>
            !
          </Typography>
          <Typography variant="h3/regular">
            You have&nbsp;
            <strong>
              {user.points}
            </strong>
            &nbsp;membership points.
          </Typography>
        </div>
```

alexzhang1618 replied:
I'm going to use component="span" since the text wrapping for the other option was a little off

SheepTester reviewed (commented):
design suggestions:

![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/22133785/cd7b3e7f-aab5-4de8-ba09-9bc4b2294ec3)

you could change the padding for `.checkin` to `padding: 1.5rem` so it lines up with the link icons

there's a lot of different font sizes in the hero. I think you should make all the smaller font sizes in the hero the same size for consistency. the smaller the better, because having text too big can feel claustrophobic and ruin the hierarchy (people might skim over it like they're headings or ads, and it can be hard to read)

I'd also reduce the line height of the big "Welcome to ACM" text, as well as increasing the spacing between that and the membership points

Also I think you mixed up the store and leaderboard icons

alexzhang1618 committed:
Address comments

alexzhang1618 committed:
Scale down mobile view

nishantbalaji commented:
Welcome text / fonts on mobile are really big, pushing more crucial info down into a scroll. Ideally we can reduce size here. Having ACM and the user's name on separate lines looks kinda awkward. 


![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/44332326/7d9f2aec-ef5f-4c2f-ab10-30ba40585cb2)

alexzhang1618 committed:
Refactor hero graphic

alexzhang1618 committed:
Move checkin to server-side, add live updating of points

alexzhang1618 committed:
Add a comment

alexzhang1618 commented:
Since the last review:
- Addressed all the visual/quality feedback
- Moved the query param check-in to GSSR
  - clear query params using window.history.replaceState
- Added some code to update the point count and attendance w/o API calls after filling out the check-in form
- Removed About Us from the navbar and instead added a link on the hero

alexzhang1618 committed:
Clear query params client side

alexzhang1618 committed:
Add a comment

alexzhang1618 committed:
Fix CSS bug, cleanup

farisashai commented on src/components/home/Hero/index.tsx:
```diff
@@ -0,0 +1,109 @@
+              <Typography
+                variant="display/heavy/medium"
+                className={styles.heading}
+                component="span"
```
do we have the ability to render a link directly with the typography component (no special styling just pass additional href)? I feel like that would help

alexzhang1618 replied:
I'll add a little enhancement to Typography so we can do 
```tsx
<Typography variant="display/heavy/medium" className={styles.heading} component="a" href={config.profileRoute} />
```

farisashai commented on src/components/home/Hero/style.module.scss:
```diff
@@ -0,0 +1,148 @@
+
+@media screen and (width <= vars.$breakpoint-md) {
+  .subheading {
+    font-size: 1rem !important;
```
maybe future issue but this needs a better solution

SheepTester replied:
I think it's one of the downsides of the `Typography` component, that it's hard to define multiple sizes for different screen sizes. Perhaps ideally, `Typography` should use globally defined classes instead of setting `style`, so it can be more easily overridden, and the CSS file can use media queries to define different font sizes for different screen sizes

farisashai commented on src/pages/index.tsx:
```diff
@@ -1,28 +1,94 @@
+    const response = await attendEvent({ token, attendanceCode });
+    const event = processCheckInResponse(response);
+    if (event) {
+      // This client side code prevents us from having to refetch user and attendance from the API.
```
just to confirm you're doing all of these on checkin

- user's membership points are updated live on the page
- users attendance is updated for the event card below
- user cookie is updated so membership points are accurate on all future pages

alexzhang1618 replied:
Membership and attendance are being updated, I'm not updating the cookie since I think we agreed on just pulling from the API to fetch user info like membership points for a page (if we rely on the cookie then across different devices it won't be accurate)

alexzhang1618 replied:
This is mainly just a little piece so info is visually accurate right after checking in to an event without having to re-call the API. If we traveled to another page where we needed the info, I'd assume we'd re-fetch

farisashai commented on src/pages/index.tsx:
```diff
@@ -49,11 +115,27 @@ const PortalHomePage = ({ pastEvents, upcomingEvents, liveEvents, attendances }:
+  let checkInResponse = null;
+  if (typeof query.code === 'string') {
+    // If a check-in code is specified, first check in to that event.
+    checkInResponse = await attendEvent({ token: authToken, attendanceCode: query.code });
```
Can you wrap this API call in an async function that returns undefined if the code query parameters isn't given so it can be batched together with the other 3 api calls

alexzhang1618 replied:
I can do that, but the intention is to run the check-in API code first since the other API calls (user + attendance) may be different depending on the outcome of the check-in API call

farisashai commented on src/pages/index.tsx:
```diff
@@ -1,28 +1,94 @@
+  const [attendance, setAttendance] = useState<PublicAttendance[]>(attendances);
+
+  const checkin = async (token: string, attendanceCode: string): Promise<void> => {
+    const response = await attendEvent({ token, attendanceCode });
```
the auth token can be fetched from the cookie client side in this function where it's used directly so we don't have to drill a ton of props through the component

farisashai commented on src/pages/index.tsx:
```diff
@@ -1,28 +1,94 @@
+    if (event) {
+      // This client side code prevents us from having to refetch user and attendance from the API.
+      setPoints(p => p + event.pointValue);
+      const a: PublicAttendance = {
```
rename to something better or inline your usage

farisashai commented on src/pages/index.tsx:
```diff
@@ -1,28 +1,94 @@
+    showToast(title, subtitle);
+    return response;
+  }
+  showToast('Unable to checkin!', response.message);
```
I thought this code was moved to the EventManager ? This feels a lot more confusing and overcomplicated separated from the actual checkin function and having to check whether we're in a success state or failure state directly in the same function

alexzhang1618 replied:
Since we also do a checkin server side, we can't use showToast in the EventManager if I'm using it in both places

alexzhang1618 committed:
Add href to Typography, fetch cookie client side

farisashai reviewed (approved):
if the typography component is specified as a "a", we should actually be rendering a `<Link>` component to handle client side routing for us but otherwise changes look good

farisashai committed:
Merge branch 'main' into feature/main-page-hero

SheepTester commented on src/components/home/Hero/style.module.scss:
```diff
@@ -0,0 +1,148 @@
+
+.inline {
+  display: inline;
+}
```
don't think this is needed anymore

SheepTester commented on src/components/home/Hero/style.module.scss:
```diff
@@ -0,0 +1,148 @@
+      gap: 0.5rem;
+
+      .heading {
+        display: inline;
```
nor is this, because you're using `component="span"` now

alexzhang1618 committed:
Fix Link stuff

alexzhang1618 committed:
Merge branch 'feature/main-page-hero' of https://github.com/acmucsd/membership-portal-ui-v2 into feature/main-page-hero

alexzhang1618 committed:
Add media query for small screens

farisashai commented on src/components/common/Typography/index.tsx:
```diff
@@ -232,6 +233,21 @@ const variantToCSS = (variant: Variant): CSSProperties => {
   const { variant, component, style, children, ...restProps } = props;
 
+  if (restProps.href) {
+    return (
```
why don't you just say Component = Link instead of rewriting the whole thing

alexzhang1618 replied:
Typescript complains because href isn't guaranteed to be true and link needs an href
